### PR TITLE
Allow Nextflow tests workflow to self-fix with `autofixable` label

### PR DESCRIPTION
**Pull request by nwiltsie**
# Description
This upgrades the Nextflow configuration regression test workflow to be way more user-friendly. You can see a demonstration of the full capabilities in https://github.com/uclahs-cds/user-nwiltsie-pipeline/pull/4.

This can be a drop-in replacement for the existing workflow, so there should be no versioning concerns.

The new behavior is enabled by labeling pull requests with `autofixable`. I've added an `autofixable` label to the [organizational defaults](https://github.com/organizations/uclahs-cds/settings/repository-defaults) so _new_ repos should inherit that, but we'll need to create that label for existing pipelines.

## User-Facing Flow for Broken Tests

If the pull request is not labeled, the workflow requests changes on the pull request:

Screenshot 2024-03-15 at 12 15 04 PM

If the pull request _is_ labeled, the workflow commits all of the re-generated test files and pushes them up. Ordinarily GitHub doesn't allow workflows to recurse, so _that_ commit is unable to be tested. To get around that, the workflow adds a trailing newline to one of the test files, then requests changes with the suggested fix of removing it:

Screenshot 2024-03-15 at 12 25 29 PM

Once someone either accepts that change _or_ pushes up a new commit, the workflow will dismiss the previous reviews:

Screenshot 2024-03-15 at 12 15 57 PM

## Weird Cases
A quirk of GitHub workflows is that they run on the _merge_ commit, not on the source branch commit. That's good, and what we want, but it can lead to strange cases like https://github.com/uclahs-cds/pipeline-call-gSV/pull/130 (specifically [this commit](https://github.com/uclahs-cds/pipeline-call-gSV/tree/ea19d9986a1d2610618dfa43b492e82189a3ed67)) where tests that did not exist in the branch were failing.

If any failing tests differ between the source and target branches, the workflow will request the following changes:

Screenshot 2024-03-15 at 12 15 16 PM

## Errata

* The workflow should not repeatedly spam the same comment - it will only issue a new review when something changes.

* The current workflows will not automatically re-run in response to the user applying the `autofixable` label - you'll need to trigger it in some other way (either by pushing up a new commit or by manually re-running the old one). I've added documentation about a second workflow that _will_ listen to label events and re-trigger the last failed run.

## Requested Feedback

From the Nextflow pipeline developers: does this solve the rough edges of updating the tests? Is there any way to make it better?

From the software team: I've tried to flush out all of the weird edge cases (see https://github.com/uclahs-cds/user-nwiltsie/pull/8 for _most_ of my debugging) but inevitably I've missed some. What can you see that I've missed?

# Checklist

- [x] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [x] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.

- [x] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [x] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

### CODE REVIEW

**Comment by yashpatel6** [comment]
Very high-level comment, rather than always automatically updating tests, would it make sense to instead provide the _option_ (through something like the way dependabot allows for rebase with a comment) to automatically update the tests? I'm thinking of cases where the tests fail and the failure was a result of an actual failure in the pipeline/config that needed to be fixed rather than the tests needing an update. That would also make the PR author have to explicitly request the update after observing a failure and checking over the work rather than the reverse of automatically fix and then remind the author to make sure changes are reviewed

**Comment by nwiltsie** [comment]
As far as I can tell, dependabot is special in that it can listen for @dependabot messages. I could not figure out how to replicate it (someone please prove me wrong, I would like that a lot more).

Assuming that we can't do "@bltestbot, fix this please", we're limited to listening to [workflow triggering events](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows). I looked at https://github.com/marketplace/actions/slash-command-dispatch for inspiration, but as far as I can tell setting up a workflow (especially a reusable workflow) using `issue_comment` will trigger on literally every comment on every pull request, even if we make them conditional. That felt super messy.

I also considered opening up another PR to be merged into the existing PR, but that felt like way too much overhead.

TL/DR: I agree, I would like a flow more like...

USER: Here are some changes
BOT: They don't work. Would you like me to fix them?
USER: Yes
BOT: Done!

... but the ability to listen for that "yes" is where I got stuck.

**Comment by yashpatel6** [comment]
Ah that's fair, I think there are ways to create custom apps/bots that would behave like dependabot but would require a fair amount of work to set up, host, maintain, etc.

Trying to have more PRs definitely seems like too much overhead, agreed

With the current setup, the autofixes could be reverted with something like: tests break -> autofix -> fix pipeline/config -> "fixed" tests break again -> autofix fixes back to original tests

**Comment by nwiltsie** [comment]
I'm aiming for the Goldilocks amount of friction in the process.

Running through a bunch of identical dependabot PRs this morning, it felt silly to keep updating the same expected changes over and over and over - it would have been nice if the Action could have just done it for me and summarized the changes for my approval. On the other hand, if it's _too_ easy, I (and others) will just get used to skimming over the changes and miss relevant items as everything happens in the background.

I'd love to be able to use the "suggest changes" feature for everything, but that is specifically limited to +/- 3 lines around existing changes:

https://github.com/github/roadmap/issues/347
https://github.com/orgs/community/discussions/4452

So... I'm completely receptive to the notion that this removes _too much_ friction and hides the real changes behind a toothless "you should review my work" comment. Any ideas for how to dial that friction level in any more?

**Comment by sorelfitzgibbon** [comment]
To summarize discussion with @nwiltsie who clarified many things for me, I lean toward the side of making it easy for the developer.  I think warnings and @nwiltsie's clear descriptions for non-experts within the PRs would be sufficient. Most developers would look carefully at the changes, and any important changes should be additionally be caught by the NFTest run which would require the developer to manually adjust the test files.
